cpu/cc2538/dev/spi.c

```diff
@@ -41,102 +41,337 @@
+#endif
+#ifdef SPI_MOSI_PIN
+#define SPI1_TX_PIN           SPI_MOSI_PIN
+#endif
+#ifdef SPI_MISO_PORT
+#define SPI1_RX_PORT          SPI_MISO_PORT
+#endif
+#ifdef SPI_MISO_PIN
+#define SPI1_RX_PIN           SPI_MISO_PIN
+#endif
+#else
+#error "Invalid SPI instance. Valid values are 0 or 1"
+#endif
+/*---------------------------------------------------------------------------*/
+/* Check port / pin settings for SPI0 and provide default values for spi_cfg */
+#ifndef SPI0_CLK_PORT
+#define SPI0_CLK_PORT         (-1)
+#endif
+#ifndef SPI0_CLK_PIN
+#define SPI0_CLK_PIN          (-1)
+#endif
+#if SPI0_CLK_PORT >= 0 && SPI0_CLK_PIN < 0 || \
+  SPI0_CLK_PORT < 0 && SPI0_CLK_PIN >= 0
+#error "Both SPI0_CLK_PORT and SPI0_CLK_PIN must be valid or invalid"
+#endif
 
-#define SPI_CLK_PORT_BASE        GPIO_PORT_TO_BASE(SPI_CLK_PORT)
-#define SPI_CLK_PIN_MASK         GPIO_PIN_MASK(SPI_CLK_PIN)
-#define SPI_MOSI_PORT_BASE       GPIO_PORT_TO_BASE(SPI_MOSI_PORT)
-#define SPI_MOSI_PIN_MASK        GPIO_PIN_MASK(SPI_MOSI_PIN)
-#define SPI_MISO_PORT_BASE       GPIO_PORT_TO_BASE(SPI_MISO_PORT)
-#define SPI_MISO_PIN_MASK        GPIO_PIN_MASK(SPI_MISO_PIN)
+#ifndef SPI0_TX_PORT
+#define SPI0_TX_PORT          (-1)
+#endif
+#ifndef SPI0_TX_PIN
+#define SPI0_TX_PIN           (-1)
+#endif
+#if SPI0_TX_PORT >= 0 && SPI0_TX_PIN < 0 || \
+  SPI0_TX_PORT < 0 && SPI0_TX_PIN >= 0
+#error "Both SPI0_TX_PORT and SPI0_TX_PIN must be valid or invalid"
+#endif
 
-/**
- * \brief Initialize the SPI bus.
- *
- * This SPI init() function uses the following defines to set the pins:
- *    SPI_CLK_PORT               SPI_CLK_PIN
- *    SPI_MOSI_PORT              SPI_MOSI_PIN
- *    SPI_MISO_PORT              SPI_MISO_PIN
- *
- * This sets the mode to Motorola SPI with the following format options:
- *    Clock phase:               1; data captured on second (rising) edge
- *    Clock polarity:            1; clock is high when idle
- *    Data size:                 8 bits
- */
+#ifndef SPI0_RX_PORT
+#define SPI0_RX_PORT          (-1)
+#endif
+#ifndef SPI0_RX_PIN
+#define SPI0_RX_PIN           (-1)
+#endif
+#if SPI0_RX_PORT >= 0 && SPI0_RX_PIN < 0 || \
+  SPI0_RX_PORT < 0 && SPI0_RX_PIN >= 0
+#error "Both SPI0_RX_PORT and SPI0_RX_PIN must be valid or invalid"
+#endif
+
+/* Here we check that either all or none of the ports are defined. As
+   we did already check that both ports + pins are either defined or
+   not for every pin, this means that we can check for an incomplete
+   configuration by only looking at the port defines */
+#if (SPI0_CLK_PORT < 0 && SPI0_TX_PORT < 0 && SPI0_RX_PORT >= 0) || \
+  (SPI0_CLK_PORT < 0 && SPI0_TX_PORT >= 0 && SPI0_RX_PORT < 0) || \
+  (SPI0_CLK_PORT < 0 && SPI0_TX_PORT >= 0 && SPI0_RX_PORT >= 0) || \
+  (SPI0_CLK_PORT >= 0 && SPI0_TX_PORT < 0 && SPI0_RX_PORT < 0) || \
+  (SPI0_CLK_PORT >= 0 && SPI0_TX_PORT < 0 && SPI0_RX_PORT >= 0) || \
+  (SPI0_CLK_PORT >= 0 && SPI0_TX_PORT >= 0 && SPI0_RX_PORT < 0)
+#error "SPI0 port / pin definition incomplete"
+#endif
```

Comment:
This could be expressed more efficiently:
```
/* If some SPI0 pads are valid, */
#if SPI0_CLK_PORT >= 0 || SPI0_MOSI_PORT >= 0 || SPI0_MISO_PORT >= 0
/* but not all */
#if SPI0_CLK_PORT < 0 || SPI0_MOSI_PORT < 0 || SPI0_MISO_PORT < 0
#error Some SPI0 pad definitions are invalid
#endif
#define SPI0_PADS_VALID
#endif
```

And same for SPI1, then you can add this coherency check at the end:
```
#ifdef SPI_DEFAULT_INSTANCE
#if SPI_DEFAULT_INSTANCE == 0
#ifndef SPI0_PADS_VALID
#error SPI_DEFAULT_INSTANCE is set to SPI0, but its pads are not valid
#endif
#elif SPI_DEFAULT_INSTANCE == 1
#ifndef SPI1_PADS_VALID
#error SPI_DEFAULT_INSTANCE is set to SPI1, but its pads are not valid
#endif
#else
#error Invalid SPI_DEFAULT_INSTANCE: valid values are 0 and 1
#endif
#endif
```